### PR TITLE
FileHandle, for reading and writing, streaming and random access

### PR DESCRIPTION
**Pull request by swankjesse**
I'm hoping to get something close to the underlying file descriptor or file
handle abstraction in POSIX.

I am unsatisfied with the design of okio.Cursor and how it will work with
random-access reading and writing. I'm hoping to make this the successor to
that class, and to completely delete that class from our public API.

I am planning a follow up PR on FileSystem, FileSystem.open(Path) that
returns a FileHandle. I will also make FileSystem.source() and
FileSystem.sink() all delegate to FileSystem.open().

### CODE REVIEW

File: okio/src/commonMain/kotlin/okio/FileHandle.kt [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+   * be released with [closeInternal] once this is 0 and [closed] is true.
+   */
+  private var openStreamCount = 0
+
+  /**
+   * Removes at least 1, and up to [byteCount] bytes from this and appends them to [sink]. Returns
+   * the number of bytes read, or -1 if this file is exhausted.
+   */
+  @Throws(IOException::class)
+  abstract fun read(offset: Long = 0L, sink: Buffer, byteCount: Long): Long
```
Comment (yschimke) [inline okio/src/commonMain/kotlin/okio/FileHandle.kt:54]: Is there any alternative to Buffer here as the target that would allow for optimised operations internally within the filesystem implementation?

File: okio/src/commonMain/kotlin/okio/FileHandle.kt [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+  private var closed = false
+
+  /**
+   * Reference count of the number of open sources and sinks on this file handle. Resources should
+   * be released with [closeInternal] once this is 0 and [closed] is true.
+   */
+  private var openStreamCount = 0
+
+  /**
+   * Removes at least 1, and up to [byteCount] bytes from this and appends them to [sink]. Returns
```
Comment (yschimke) [inline okio/src/commonMain/kotlin/okio/FileHandle.kt:50]: I wonder if there is any annotation to use here that you can't ignore the return value?

File: okio/src/commonMain/kotlin/okio/FileHandle.kt [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+   * the number of bytes read, or -1 if this file is exhausted.
+   */
+  @Throws(IOException::class)
```
Comment (yschimke) [inline okio/src/commonMain/kotlin/okio/FileHandle.kt:53]: What happens with partial and likely interrupted or failed reads?

File: okio/src/commonMain/kotlin/okio/FileHandle.kt [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    override fun flush() {
+      fileHandle.flush()
+    }
+
+    override fun timeout() = Timeout.NONE
+
+    override fun close() {
+      synchronized(this) {
```
Comment (yschimke) [inline okio/src/commonMain/kotlin/okio/FileHandle.kt:185]: Is synchronizing on the sink here correct? Couldn't that race if two sinks or sources are closed at same time?

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 1 above)
Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:54]: If we wanna do such optimizations, we can change buffer to support ’em!

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 2 above)
Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:50]: Yeah, it’s `CheckReturnValue`. This is the same as the signature of Source.read().

(Follow-up?)

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 3 above)
Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:53]: Yeah, we’re already potentially problematic here.

Tracking as: https://github.com/square/okio/issues/917

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 4 above)
Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt]: Totally incorrect. Great catch.

Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt]: Fixed to synchronize on `fileHandle`.

Great catch. Hooray for your careful code review.

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 1 above)
Comment (yschimke) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:54]: I was wondering what a file to file direct copy would look like? Operations like https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileChannel.html#transferFrom(java.nio.channels.ReadableByteChannel,%20long,%20long)

Comment (yschimke) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:54]: I wonder whether there is a good set of common operations to support? https://www.baeldung.com/java-filechannel

File: okio/src/commonMain/kotlin/okio/FileHandle.kt [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    return FileHandleSource(this, offset)
+  }
+
+  /**
+   * Returns the position of [source] in the file. The argument [source] must be either a source
+   * produced by this file handle, or a [BufferedSource] that directly wraps such a source. If the
+   * parameter is a [BufferedSource], it adjusts for buffered bytes.
+   */
+  @Throws(IOException::class)
+  fun position(source: Source): Long {
```
Comment (bnorm) [inline okio/src/commonMain/kotlin/okio/FileHandle.kt:89]: What's the benefit of this API over exposing `FileSystemSource` as a public type with a public `position` property? This could get awkward if doing more than buffering the source/sink, for example, hashing to validate a checksum.

File: okio/src/commonMain/kotlin/okio/FileHandle.kt [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    var source = source
+    var bufferSize = 0L
+
+    if (source is RealBufferedSource) {
```
Comment (bnorm) [inline okio/src/commonMain/kotlin/okio/FileHandle.kt:93]: There's no protection from double-buffering a source or sink. Should this be a `while`? Or should that protection be added in a separate PR?

File: okio/src/commonMain/kotlin/okio/FileHandle.kt [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+  /**
+   * Returns the total number of bytes in the file. This will change if the file size changes.
+   */
+  @Throws(IOException::class)
+  abstract fun size(): Long
+
+  /** Removes [byteCount] bytes from [source] and writes them to this at [offset]. */
+  @Throws(IOException::class)
+  abstract fun write(offset: Long = 0L, source: Buffer, byteCount: Long)
```
Comment (bnorm) [inline okio/src/commonMain/kotlin/okio/FileHandle.kt:64]: Is it going to be common enough to repeatedly read from/write to the beginning of a file such that there is a benefit to having a default value for offset? Would it be better without to avoid any confusion between this and Source/Sink style reading/writing? It's probably pretty low risk though so this might be bikeshedding territory.

**Review by oldergod — APPROVED** [review]

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 1 above)
Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:54]: That’s a good list. I think probably these:

- Reading
- Writing
- Current Position
- Get file size
- Set file size (ie. truncate)
- Force data
- Force metadata

And not these:
 - Memory mapping
 - Locks

These last two are exotic enough that omitting them shouldn’t be missed by most developers.

Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:54]: As for direct filesystem transfer, that’s a neat optimization and it’s possible our APIs don’t preclude them!

But it’s a difficult API to use because it’s easy to fall off the fast path, and there’s no way to make sure that you’re on the fast path.

I’d prefer to omit that.

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 5 above)
Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:89]: Yeah, you’re absolutely right. I definitely wrote that out and tried to see how it’d play out. The biggest drawback of it is how it interacts with `BufferedSource`.

If we introduce a new type then `BufferedSource` either won’t implement the interface, or other `BufferedSource` implementations will implement the interface but not be able to do anything reasonable in this method.

The use case I’m imagining for this is passing a BufferedSource to a method to read some data, then checking the new position.

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 6 above)
Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:93]: I contemplated that too. I think this API is too fragile to be used dynamically like that. This API isn’t so flexible to take an arbitrary transformation of a source; just the source it created and a direct wrapping of it.

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 7 above)
Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:64]: Great point. I think you’re right; the default here doesn’t make sense. Removed!

File: okio/src/commonMain/kotlin/okio/FileHandle.kt [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+  /**
+   * Subclasses should implement this to release resources held by this file handle. It is invoked
+   * once both the file handle is closed, and also all sources and sinks produced by it are also
+   * closed.
+   */
+  @Throws(IOException::class)
+  protected abstract fun closeInternal()
+
+  private class FileHandleSink(
```
Comment (yschimke) [inline okio/src/commonMain/kotlin/okio/FileHandle.kt:167]: "inner class" skip the val constructor param?

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 6 above)
Comment (yschimke) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:93]: It's already really brittle here, the type system is not protecting you here.  So it pretty much needs to be the source that you immediately called buffer on and never let out of your sight.

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 8 above)
Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:167]: I will, if you can help me figure out how to make this work in the enclosing class:

```
    require(source is FileHandleSource && source.fileHandle === this) {
      "source was not created by this FileHandle"
    }
```

Comment (oldergod) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:167]: It could work if you did something like this I think.
```kotlin
  inner class FileHandleSink(
    var position: Long
  ) : Sink {
    var fileHandle = this@FileHandle
```

But I don't care either way how you end up writing it. I like constructor param.

Comment (yschimke) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:167]: [https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS40LjMwIiwicGxhdGZvcm0iOiJqYXZhIiwiYXJncyI6IiIsImpzQ29kZSI6IiIsIm5vbmVNYXJrZXJzIjp0cnVlLCJ0aGVtZSI6ImlkZWEiLCJjb2RlIjoiY2xhc3MgQSB7XG4gICAgZnVuIGIoKSA9IEIoKVxuICAgIFxuICAgIGlubmVyIGNsYXNzIEIge1xuICAgICAgICBmdW4gcnVuKCkge1xuICAgICAgICAgICAgcHJpbnRsbih0aGlzKVxuICAgICAgICAgICAgcHJpbnRsbih0aGlzQEEpXG4gICAgICAgICAgICBwcmludGxuKHRoaXNAQilcbiAgICAgICAgfVxuICAgIH1cbn1cblxuZnVuIG1haW4oKSB7XG4gICAgQSgpLmIoKS5ydW4oKVxufSJ9](play.kotlinlang.org)

```
class A {
    fun b() = B()
    
    inner class B {
        fun run() {
            println(this)
            println(this@A)
            println(this@B)
        }
    }
}

fun main() {
    A().b().run()
}
```

Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:167]: How about this one?
https://pl.kotl.in/GiG6aSoGI

Comment (yschimke) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:167]: Yeah, your solution seems cleaner.

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 1 above)
Comment (yschimke) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt]: Maybe add another case which is dropping into native APIs.  Make sure that you design for that case.  Here is how to bridge to java.nio, or node filesystem.

Comment (yschimke) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt]: To be clear, I'm not pushing for where the line is, just being clear around where it is and an image in your mind that it will generally work for typical client use cases, combined with the types of FileSystems you see implemented.

File: okio/src/commonMain/kotlin/okio/FileHandle.kt (hunk 6 above)
Comment (yschimke) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:93]: I'm now worried about this breaking with even a simple decorated file system, like logging.

Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/FileHandle.kt:93]: When it implements `position` it'll need to delegate with the unwrapped instance.

File: okio/src/commonMain/kotlin/okio/FileHandle.kt [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+  @Throws(IOException::class)
+  abstract fun flush()
+
+  /**
+   * Returns a source that reads from this starting at [offset]. The returned source must be closed
+   * when it is no longer needed.
+   */
+  @Throws(IOException::class)
+  fun source(offset: Long = 0L): Source {
+    synchronized(this) {
```
Comment (JakeWharton) [inline okio/src/commonMain/kotlin/okio/FileHandle.kt:76]: what in the magic hell synchronized is this that it works (correctly) from common code? shouldn't we be using atomicfu?

File: okio/src/commonMain/kotlin/okio/FileHandle.kt [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+  /** Pushes all buffered bytes to their final destination. */
+  @Throws(IOException::class)
+  abstract fun flush()
+
+  /**
+   * Returns a source that reads from this starting at [offset]. The returned source must be closed
+   * when it is no longer needed.
+   */
+  @Throws(IOException::class)
+  fun source(offset: Long = 0L): Source {
```
Comment (JakeWharton) [inline okio/src/commonMain/kotlin/okio/FileHandle.kt:75]: jvm overloads? just in case!

File: okio/src/commonMain/kotlin/okio/FileHandle.kt [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+    return source.position - bufferSize
+  }
+
+  /**
+   * Returns a sink that writes to this starting at [offset]. The returned sink must be closed when
+   * it is no longer needed.
+   */
+  @Throws(IOException::class)
+  fun sink(offset: Long = 0L): Sink {
```
Comment (JakeWharton) [inline okio/src/commonMain/kotlin/okio/FileHandle.kt:110]: jvm overloads? just in case!

**Review by JakeWharton — COMMENTED** [review]
👍
